README
NAME

  Latin2 - Source code filter to escape Latin-2 script

SYNOPSIS

  use Latin2; # usage for application programmers

DESCRIPTION

  The Latin2 software provides character oriented Perl scripting
  environment in Latin-2 by easy installation and easy usage.

  - Character oriented regular expression
  - Character oriented runtime routines
  - Character oriented subroutines
      and
  - Byte oriented CORE::* functions
  - Byte oriented regular expression on /b modifier

  Information processing model beginning with Perl3 or this software.

    +--------------------------------------------+
    |       Text strings as Binary strings       |
    |       Binary strings as Text strings       |
    +--------------------------------------------+
    |        Not UTF8 Flagged, UNIX/C-ism        |
    +--------------------------------------------+

  Script could be written in native encoding of operating systems.
  - Like contents of a file
  - Like a file name on the file systems
  - Like command lines
  - Like environment variables
  - Like parameters of API

  In UNIX Everything is a File
  - In UNIX everything is a stream of bytes
  - In UNIX the filesystem is used as a universal name space

INSTALLATION

  Just copy Latin2.pm and Elatin2.pm to your @INC directory.
  For example, to C:\Perl\site\lib

SUBROUTINES

  Latin2::eval(charstring)
  Latin2::length(charstring)
  Latin2::substr(charstring, charpos, charlen)
  and more...

SEE ALSO

  http://search.cpan.org/~ina/



Changes
Revision history for this software.

1.02  2015-01-01 00:00:00
  - support Sjis::eval
  - support StrawberryPerl 5.20.1
  - support ActivePerl 5.20.1
  - optimize LONG_STRING_FOR_RE
  - created by INABA Hitoshi

1.01  2014-09-08 00:00:00
  - avoid "Octal number in vector unsupported" on perl 5.6
  - fix broken t/243_filetest.t of Big5Plus
  - created by INABA Hitoshi

1.00  2014-09-01 00:00:00
  - support EUC-TW
  - support yada-yada or triple-dot operator
  - created by INABA Hitoshi

0.99  2014-08-03 00:00:00
  - fix fix name: in META.yml Char::Sjis to Char-Sjis
  - fix sub e_capture() on SBCS
  - created by INABA Hitoshi

0.98  2014-08-02 00:00:00
  - experimental release
  - created by INABA Hitoshi

0.97  2014-07-28 00:00:00
  - fix name: in META.yml
  - support Perl 5.20.0
  - support StrawberryPerl 5.20.0
  - support perl on Raspberry Pi
  - created by INABA Hitoshi

0.96  2014-04-26 00:00:00
  - fix m/[ABC]/i
  - created by INABA Hitoshi

0.95  2014-01-21 00:00:00
  - fix using "use 5.00503;"
  - created by INABA Hitoshi

0.94  2014-01-13 00:00:00
  - support CaseFolding-6.3.0.txt
  - support Perl 5.18.2
  - created by INABA Hitoshi

0.93  2013-12-29 00:00:00
  - fix quoting after filetest operators
  - Esjis::unlink() deletes read only file too
  - remove PERL5.BATes
  - remove strict.pm_, warnings.pm_, warnings/register.pm_, and feature.pm_
  - truncate(), getpwnam(), and getpwuid() into eval q{}
  - consider Win95Cmd.exe by PERL5SHELL
  - remove $ENV{'COMSPEC'}
  - changed condition for escaping script
  - support ActivePerl 5.18.1.1800 Windows (x86)
  - support ActivePerl 5.18.1.1800 Windows (64-bit, x64)
  - created by INABA Hitoshi

0.92  2013-07-18 00:00:00
  - support JIS8
  - support Arabic
  - support Hebrew
  - support TIS-620
  - created by INABA Hitoshi

0.91  2013-07-07 00:00:00
  - improve split() for over 32766 octet string
  - add short README
  - use Esjis ver.sion; --> use Esjis ver.sion.0;
  - undo BEGIN { unshift @INC, '/Perl/site/lib/Sjis'; } in Esjis.pm
  - created by INABA Hitoshi

0.90  2013-06-08 00:00:00
  - support Perl 5.18.0
  - support StrawberryPerl 5.18.0
  - \s in a regex matches a vertical tab on perl 5.18.0 or later
  - support lvalue Sjis::substr()
  - follow File::HomeDir 1.00
  - support DWIM Perl
  - add dperl.bat
  - created by INABA Hitoshi

0.89  2013-05-12 00:00:00
  - cancel improving multi-byte anchoring for over 32766 octet string on perl 5.10.0
  - add perl518.bat, perl64518.bat, aperl518.bat, aperl64518.bat, sperl518.bat, and sperl64518.bat
  - support Cygwin's Perl
  - add cperl.bat
  - created by INABA Hitoshi

0.88  2013-05-01 00:00:00
  - improve multi-byte anchoring for over 32766 octet string
  - remove multi-byte anchoring of UTF2 and OldUTF8
  - support ActivePerl 5.16.3
  - support ActivePerl 5.14.4
  - fix errata
  - created by INABA Hitoshi

0.87  2013-03-22 00:00:00
  - support Strawberry Perl 5.16.3
  - support Perl 5.16.3
  - support Strawberry Perl 5.14.4
  - support Perl 5.14.4
  - fix errata
  - remove module (Win32.pm)
  - created by INABA Hitoshi

0.86  2013-01-27 00:00:00
  - fix Sjis::getc(BAREWORD)
  - created by INABA Hitoshi

0.85  2013-01-13 00:00:00
  - support Microsoft Windows 95
  - support Microsoft Windows NT 4.0
  - support KPS 9566
  - support Sjis::getc()
  - created by INABA Hitoshi

0.84  2012-11-19 00:00:00
  - support Strawberry Perl 5.16.2
  - support Perl 5.16.2
  - delete escape-characters of run-time command
  - have to use both two-argument and three-argument open()
  - support Microsoft Windows 8 Pro (x64)
  - support PowerShell 3.0
  - created by INABA Hitoshi

0.83  2012-09-29 00:00:00
  - support Strawberry Perl 5.16.1
  - support ActivePerl 5.16.1
  - support x64 of ActivePerl and Strawberry Perl
  - new ver.sion.0 check (match or die)
  - fix $_ !~ s/regexp/replacement/
  - remove ESCAPE_SECOND_OCTET on single byte encoding
  - escape left brace on un-escaping of \N{...}, \p{...}, and \P{...}
  - remove "distribution_type: script" from META.yml
  - add perl64512.bat, perl64514.bat, perl64516.bat, aperl64512.bat, aperl64514.bat, aperl64516.bat, sperl64512.bat, sperl64514.bat, and sperl64516.bat
  - created by INABA Hitoshi

0.82  2012-07-29 00:00:00
  - support Strawberry Perl 5.16.0
  - support ActivePerl 5.16.0
  - support variables in character class in regular expression
  - created by INABA Hitoshi

0.81  2012-05-25 00:00:00
  - support fold case escape \F and function fc()
  - support scalar context of Esjis::stat, Esjis::stat_, Esjis::lstat, and Esjis::lstat_
  - support Strawberry Perl 5.8, 5.10, 5.12, and 5.14
  - become independent of strict.pm
  - add feature.pm
  - add perl5.bat
  - add aperl58.bat, aperl510.bat, aperl512.bat, aperl514.bat, and aperl516.bat
  - add sperl58.bat, sperl510.bat, sperl512.bat, sperl514.bat, and sperl516.bat
  - fix dummy access (-T and -B for underline cache) in Esjis::T, Esjis::T_, Esjis::B, and Esjis::B_
  - fix $dbh->Esjis::do to $dbh->do on DBI
  - fix Esjis::split(/(-)|(,)/,"1-10,20") --> (1,"-",undef,10,undef,",",20)
  - fix "\v" means "\x0A", "\x0B", "\x0C", and "\x0D"
  - remove './' at head of return value from Esjis::glob and Esjis::glob_
  - created by INABA Hitoshi

0.80  2012-03-22 00:00:00
  - support byte-oriented regexp //b, m//b, m''b, s///b, s'''b, qr//b, qr''b, split(//b), split(m//b), split(m''b), split(qr//b), and split(qr''b)
  - support match once only ??, m??, ??b, and m??b
  - support literal null string pattern
  - support $`, $PREMATCH, ${^PREMATCH}, $&, $MATCH, ${^MATCH}, &', $POSTMATCH, and ${^POSTMATCH}
  - add perl516.bat
  - created by INABA Hitoshi

0.79  2011-12-23 00:00:00
  - fix m/[[:lower:]]/i, m/[[:upper:]]/i, m/[[:^lower:]]/i, and m/[[:^upper:]]/i
  - created by INABA Hitoshi

0.78  2011-10-28 00:00:00
  - support Windows-1258
  - update documents and URL
  - created by INABA Hitoshi

0.77  2011-07-27 00:00:00
  - support KOI8-R, KOI8-U, US-ASCII
  - created by INABA Hitoshi

0.76  2011-06-30 00:00:00

test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

use Latin2;

$| = 1;
print "1..1\n";
print "ok - 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):



pMakefile.PL
######################################################################
#
# pMakefile.PL - Makefile.PL of Perl Poor Tools
#
# Copyright (c) 2008, 2009, 2010, 2011, 2014, 2015 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;

# make pmake.bat
if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
    open(FH_MAKEBAT, '>pmake.bat') || die "Can't open file: pmake.bat\n";
    print FH_MAKEBAT <<'END';
@rem = '--*-Perl-*--
@echo off
if "%OS%" == "Windows_NT" goto WinNT
perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
goto endofperl
:WinNT
perl -x -S "%0" %*
if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
if %errorlevel% == 9009 echo You do not have Perl in your PATH.
goto endofperl
@rem ';
#!perl
#line 14
undef @rem;
END
    print FH_MAKEBAT <DATA>;
    close FH_MAKEBAT;
}

# make pmake
else {
    open(FH_MAKEBAT, '>pmake') || die "Can't open file: pmake\n";
    print FH_MAKEBAT '#!', &which($^X), "\n";
    print FH_MAKEBAT <DATA>;
    close FH_MAKEBAT;
    chmod 0755, 'pmake';
}

sub which {
    if ($_[0] =~ m#\A / #oxms) {
        return $_[0];
    }
    else {
        for my $path (split(/:/,$ENV{'PATH'})) {
            if (-e qq{$path/$_[0]}) {
                return qq{$path/$_[0]};
            }
        }
        return $_[0];
    }
}

__END__
######################################################################
#
# pmake - make of Perl Poor Tools
#
# Copyright (c) 2008, 2009, 2010 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;
use File::Path;
use File::Copy;
use File::Basename;
use Test::Harness;

unless (@ARGV) {
    if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
        die <<END;

usage: pmake
usage: pmake test
usage: pmake install
usage: pmake dist
usage: pmake ptar.bat
usage: pmake pwget.bat

END
    }
    else {
        die <<END;

usage: pmake
usage: pmake test
usage: pmake install
usage: pmake dist
usage: pmake ptar
usage: pmake pwget

END
    }
}

# get file list
open(FH_MANIFEST, 'MANIFEST');
chomp(my @file = <FH_MANIFEST>);
close FH_MANIFEST;

for my $target (@ARGV) {
    if ($target eq 'test') {
        my @test = grep m{ \A (?: test\.pl | t/.+\.t ) \z }xmsi, @file;
        runtests(@test);
    }
    elsif ($target eq 'install') {

        # install *.pm files to /Perl/site/lib
        my $perl_site_lib = '';
        if (($perl_site_lib) = grep(m{site_perl}xms, reverse @INC)) {
        }
        elsif (($perl_site_lib) = grep(m{site}xms, reverse @INC)) {
        }
        if ($perl_site_lib ne '') {
            for (grep m/ \. pm \z /xmsi, @file) {
                if (m#^(.+)/#) {
                    print STDERR "copy $_ $perl_site_lib/$1\n";
                    mkpath("$perl_site_lib/$1", 0, 0777) unless -d "$perl_site_lib/$1";
                    copy($_, "$perl_site_lib/$1");
                }
                else {
                    print STDERR "copy $_ $perl_site_lib\n";
                    copy($_, $perl_site_lib);
                }
            }
        }

        # install *.pl, *.bat, *.exe, and *.com files to /Perl/bin
        my($perl_bin) = $^X =~ /^(.*)\\[^\\]*$/;
        for (grep m/ \. (?: pl | bat | exe | com ) \z /xmsi, @file) {
            next if m/(?: Makefile | pMakefile | test ) \.pl  $/xmsi;
            next if m/(?: pmake | ptar )                \.bat $/xmsi;
            if (m#^(.+)/#) {
                print STDERR "copy $_ $perl_bin/$1\n";
                mkpath("$perl_bin/$1", 0, 0777) unless -d "$perl_bin/$1";
                copy($_, "$perl_bin/$1");
            }
            else {
                print STDERR "copy $_ $perl_bin\n";
                copy($_, $perl_bin);
            }
        }
    }
    elsif ($target eq 'dist') {

        # make work directory
        my($changes) = grep(/^Changes$/i,@file);
        open(FH_CHANGES, $changes) || die "Can't open file: $changes\n";
        my $version = 0;
        while (<FH_CHANGES>) {
            if (/^(\d+\.\d+(\.\d+\.\d+)*)/) {
                if ($1 > $version) {
                    $version = $1;
                }
            }
        }
        close FH_CHANGES;
        my $dirname  = dirname($file[0]);
        my $basename = basename($file[0], '.pm','.pl','.bat');
        my $tardir;
        if ($dirname eq 'lib') {
            $tardir = "$basename-$version";
        }
        elsif ($dirname ne '.') {
            $dirname =~ tr#/#-#;
            $tardir = "$dirname-$basename-$version";
        }
        else {
            $tardir = "$basename-$version";
        }
        rmtree($tardir,0,0);

        # rewrite META.yml file
        if (my($metayml) = grep(/^META\.yml$/i,@file)) {
            if (open(FH_METAYML, $metayml)) {
                $_ = join('',<FH_METAYML>);
                close FH_METAYML;
                if (open(FH_METAYML, ">$metayml")) {
                    s/^(version:\s*)\d+\.\d+(\.\d+\.\d+)*/$1$version/m;
                    binmode FH_METAYML;
                    print FH_METAYML $_;
                    close FH_METAYML;
                }
            }
        }

        if ($^O =~ /solaris|linux/i) {
            system(qq{tar -cvf $tardir.tar $tardir});
            system(qq{gzip $tardir.tar});
        }
        else {
            eval q{
                use Compress::Zlib;
                use Archive::Tar;
            };

            # make *.tar file
            my $tar = Archive::Tar->new;
            for my $file (@file) {
                if (-e $file) {
                    mkpath(dirname("$tardir/$file"), 0, 0777);
                    print STDERR "copy $file $tardir/$file\n";
                    copy($file, "$tardir/$file");

#-----------------------------------------------------------------------------
# Sunday December 21, 2008 07:38 PM 
# Fixing world writable files in tarball before upload to CPAN [ #38127 ]
# http://use.perl.org/~bart/journal/38127 (dead link)
# Fix CPAN uploads for world writable files
# http://perlmonks.org/index.pl?node_id=731935
#-----------------------------------------------------------------------------
#                   $tar->add_files("$tardir/$file");
#-----------------------------------------------------------------------------
                    open(FH, $file) || die "Can't open file: $file\n";
                    binmode FH;
                    local $/ = undef; # slurp mode
                    my $data = <FH>;
                    close FH;
                    if ($file =~ m/\. (com|exe|bat|pl) \z/oxmsi) {
                        $tar->add_data("$tardir/$file", $data, {'mode' => 0775});
                    }
                    else {
                        $tar->add_data("$tardir/$file", $data, {'mode' => 0664});
                    }
#-----------------------------------------------------------------------------
                }
                else {
                    die "file: $file is not exists.\n";
                }
            }

            $tar->write("$tardir.tar");
            rmtree($tardir,0,0);

            # make *.tar.gz file
            my $gz = gzopen("$tardir.tar.gz", 'wb');
            open(FH_TAR, "$tardir.tar") || die "Can't open file: $tardir.tar\n";
            binmode FH_TAR;
            $gz->gzwrite(join('',<FH_TAR>));
            close FH_TAR;
            $gz->gzclose;
            unlink "$tardir.tar";
        }

        # P.565 Cleaning Up Your Environment
        # in Chapter 23: Security
        # of ISBN 0-596-00027-8 Programming Perl Third Edition.

        # local $ENV{'PATH'} = '.';
        local @ENV{qw(IFS CDPATH ENV BASH_ENV)};

        # untar test
        if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
            system(qq{pmake.bat ptar.bat});
            system(qq{ptar.bat xzvf $tardir.tar.gz});
        }
        else {
            system(qq{./pmake ptar});
            system(qq{./ptar xzvf $tardir.tar.gz});
        }
    }
    elsif ($target =~ /^ptar(?:\.bat)?$/) {

        my $ptar = <<'PTAR_END';
######################################################################
#
# ptar - tar of Perl Poor Tools
#
# Copyright (c) 2008, 2009, 2010, 2011 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;

if ($ARGV[0] ne 'xzvf') {
    die <<END;

usage: ptar xzvf file1.tar.gz file2.tar.gz ...

** This ptar supports xzvf option only. **

e(x)tract
(z)ip-file
(v)erbose
(f)ile

END
}

for my $gzfile (grep m/\.tar\.gz$/xmsi, @ARGV) {

    if ($^O =~ /solaris|linux/i) {
        system(qq{gzip -cd $gzfile | tar -xvf -});
    }
    else {
        eval q{
            use Compress::Zlib;
            use Archive::Tar;
        };

        my $gz = gzopen($gzfile, 'rb');
        (my $tarfile = $gzfile) =~ s/\.gz$//xmsi;
        open(FH_TAR, ">$tarfile") || die "Can't open file: $tarfile\n";
        binmode FH_TAR;
        while ($gz->gzreadline(my $line)) {
            print FH_TAR $line;
        }
        $gz->gzclose;
        close FH_TAR;

        my $tar = Archive::Tar->new($tarfile,1);
        for my $file ($tar->list_files){
            if (-e $file) {
                print STDERR "skip $file is already exists.\n";
            }
            else {
                print STDERR "x $file\n";
                $tar->extract($file);
            }
        }
        unlink $tarfile;
    }
}

__END__
PTAR_END

        # make ptar.bat
        if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
            open(FH_TARBAT, '>ptar.bat') || die "Can't open file: ptar.bat\n";
            print FH_TARBAT <<'END';
@rem = '--*-Perl-*--
@echo off
if "%OS%" == "Windows_NT" goto WinNT
perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
goto endofperl
:WinNT
perl -x -S "%0" %*
if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
if %errorlevel% == 9009 echo You do not have Perl in your PATH.
goto endofperl
@rem ';
#!perl
#line 14
undef @rem;
END
            print FH_TARBAT $ptar;
            print FH_TARBAT ":endofperl\n";
            close FH_TARBAT;
        }

        # make ptar
        else {
            open(FH_TARBAT, '>ptar') || die "Can't open file: ptar\n";
            print FH_TARBAT '#!', &which($^X), "\n";
            print FH_TARBAT $ptar;
            close FH_TARBAT;
            chmod 0755, 'ptar';
        }
    }
    elsif ($target =~ /^pwget(?:\.bat)?$/) {

        my $pwget = <<'PWGET_END';
######################################################################
#
# pwget - wget of Perl Poor Tools
#
# Copyright (c) 2011 INABA Hitoshi <ina@cpan.org>
######################################################################

use Socket;

unless (@ARGV) {
    die <<END;

usage: pwget http://www.foo.com/bar/baz.tar.gz

END
}

my $url = $ARGV[0];
my $forward = 3;
while ($forward-- > 0) {
    my($hostname) = $url =~ m#http://([^/]+)/#;
    my $port = ($hostname =~ s/:([0-9]+)//) ? $1 : 80;

    socket(SOCKET,PF_INET,SOCK_STREAM,getprotobyname('tcp')) || die "Can't open TCP/IP socket.\n";
    connect(SOCKET,sockaddr_in($port,inet_aton($hostname)))  || die "Can't connect to $hostname:$port.\n";
    select SOCKET;
    $| = 1;
    select STDOUT;

    my $request = <<END;
GET $url HTTP/1.0
Accept: */*
User-Agent: $0

END
    $request =~ s/\n/\r\n/g;
    print SOCKET $request;

    my($head,$body) = split(/\r\n\r\n/,join('',<SOCKET>),2);
    close SOCKET;

    if ($head =~ m#^Location: (\S+)#ms) {
        $url = $1;
        print STDERR "Location: $url\n";
        next;
    }

    my($file) = $ARGV[0] =~ m#([^/]+)$#;
    open(FILE,">$file") || die "Can't open file: $file\n";
    binmode FILE;
    print FILE $body;
    close FILE;
    if ($head =~ m#Content-Length: ([0-9]+)#ms) {
        if (-s $file == $1) {
            print STDERR "ok - $file\n";
        }
        else {
            print STDERR "not ok - $file\n";
            unlink $file;
        }
    }
    last;
}

__END__
PWGET_END

        # make pwget.bat
        if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
            open(FH_WGETBAT, '>pwget.bat') || die "Can't open file: pwget.bat\n";
            print FH_WGETBAT <<'END';
@rem = '--*-Perl-*--
@echo off
if "%OS%" == "Windows_NT" goto WinNT
perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
goto endofperl
:WinNT
perl -x -S "%0" %*
if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
if %errorlevel% == 9009 echo You do not have Perl in your PATH.
goto endofperl
@rem ';
#!perl
#line 14
undef @rem;
END
            print FH_WGETBAT $pwget;
            print FH_WGETBAT ":endofperl\n";
            close FH_WGETBAT;
        }

        # make pwget
        else {
            open(FH_WGETBAT, '>pwget') || die "Can't open file: pwget\n";
            print FH_WGETBAT '#!', &which($^X), "\n";
            print FH_WGETBAT $pwget;
            close FH_WGETBAT;
            chmod 0755, 'pwget';
        }
    }
    else {
        warn "unknown target: $target.\n";
    }
}

sub which {
    if ($_[0] =~ m#\A / #oxms) {
        return $_[0];
    }
    else {
        for my $path (split(/:/,$ENV{'PATH'})) {
            if (-e qq{$path/$_[0]}) {
                return qq{$path/$_[0]};
            }
        }
        return $_[0];
    }
}

__END__
:endofperl
